### PR TITLE
minio S3 documentation

### PR DESCRIPTION
**Pull request by felix-engelmann**
To have full control over your stream, you may self host the video chunks on MinIO.

### CODE REVIEW

**Comment by gabek** [comment]
Nice!

File: content/docs/S3.md [hunk 1]
```diff
@@ -186,4 +187,34 @@ You can have B2 delete old video files every day to keep your storage low by sim
 
 {{< img src="/docs/img/b2_delete_old_files.png" align="center">}}
 
+## MinIO
+
+If you want to host the video segments on a self hosted S3 compatible [MinIO](https://min.io/) server, you get even more control over your data.
+
+### Bucket
+
+Create a bucket, e.g. with the [minio client](https://docs.min.io/docs/minio-client-complete-guide.html) by
```
Comment (jeyemwey) [inline content/docs/S3.md:196]:
```suggestion
Create a bucket, e.g. with the [MinIO client](https://docs.min.io/docs/minio-client-complete-guide.html) by
```

Fix typo

File: content/docs/S3.md [hunk 2]
```diff
@@ -186,4 +187,34 @@ You can have B2 delete old video files every day to keep your storage low by sim
...
+
+* add an alias `mc alias set <ALIAS> <YOUR-S3-ENDPOINT> [YOUR-ACCESS-KEY] [YOUR-SECRET-KEY]`
+* create a new bucket `mc mb <ALIAS>/stream`
+* allow downloading from bucket `mc policy set download <ALIAS>/stream`
+
+### CORS
+
+Make sure to allow access to the buckets file from your owncast domain.
```
Comment (jeyemwey) [inline content/docs/S3.md:204]: Can you link to some documentation? I’ve seen many people having problems with CORS in the Owncast discussions, so a link to a tutorial or documentation would be great.

File: content/docs/S3.md [hunk 3]
```diff
@@ -186,4 +187,34 @@ You can have B2 delete old video files every day to keep your storage low by sim
...
 
+## MinIO
+
+If you want to host the video segments on a self hosted S3 compatible [MinIO](https://min.io/) server, you get even more control over your data.
+
+### Bucket
+
+Create a bucket, e.g. with the [minio client](https://docs.min.io/docs/minio-client-complete-guide.html) by
+
+* add an alias `mc alias set <ALIAS> <YOUR-S3-ENDPOINT> [YOUR-ACCESS-KEY] [YOUR-SECRET-KEY]`
```
Comment (jeyemwey) [inline content/docs/S3.md:198]: Who defines `[YOUR-ACCESS-KEY]` and `[YOUR-SECRET-KEY]`? Is that a random passphrase that the user can enter and has to generate somewhere else?

**Review by jeyemwey — CHANGES_REQUESTED** [review]
Hi, thank you for your awesome contribution to our documentation. :) There are few things that I’d like to change before merging, but then I’m good to go :)

File: content/docs/S3.md (hunk 3 above)
Comment (felix-engelmann) [reply content/docs/S3.md:199]: I'm not sure if I should explain the setup of MinIO itself too. Therefore I referenced an example docker-compose

File: content/docs/S3.md (hunk 2 above)
Comment (felix-engelmann) [reply content/docs/S3.md]: As it depends on the reverse proxy in front of MinIO, I explained it for traefik.

File: content/docs/S3.md (hunk 3 above)
Comment (gabek) [reply content/docs/S3.md:199]: I think this is good how it is.  If you use MinIO then you understand those values and how to use them.  If you don't already use MinIO then this documentation probably isn't for you.

File: content/docs/S3.md [hunk 4]
```diff
@@ -186,4 +187,44 @@ You can have B2 delete old video files every day to keep your storage low by sim
 
 {{< img src="/docs/img/b2_delete_old_files.png" align="center">}}
 
+## MinIO
+
+If you want to host the video segments on a self hosted S3 compatible [MinIO](https://min.io/) server, you get even more control over your data. A docker-compose deployment example is available in the [examples directory](https://github.com/owncast/owncast/blob/master/examples/docker-compose.yaml).
```
Comment (jeyemwey) [inline content/docs/S3.md:192]: I'm not sure if the normal docker compose should have a minion host. That would be another thing to test against, while this is just part of the documentation.

File: content/docs/S3.md [hunk 5]
```diff
@@ -186,4 +187,44 @@ You can have B2 delete old video files every day to keep your storage low by sim
...
+Make sure to allow access to the buckets file from your owncast domain. If your MinIO is deployed behind a traefik reverse proxy, make sure to add a middleware to your MinIO container:
+
+{{< highlight yaml >}}
+labels:
+  - "traefik.http.routers.minio.middlewares=livecors"`
+  - "traefik.http.middlewares.livecors.headers.accesscontrolallowmethods=GET,OPTIONS"
+  - "traefik.http.middlewares.livecors.headers.allowedhosts=https://s3-storage.your.org"
+{{< / highlight >}}
+  
+For details, c.f. the [documentation](https://doc.traefik.io/traefik/middlewares/headers/#cors-headers), respective of the reverse proxy you use.
```
Comment (gabek) [inline content/docs/S3.md:214]: I'd vote just to stop at saying "Make sure to allow access to the buckets file from your owncast domain".  The rest is a lot of detail, and if people don't use traefik then it's not terribly useful anyway.

File: content/docs/S3.md [hunk 6]
```diff
@@ -186,4 +187,44 @@ You can have B2 delete old video files every day to keep your storage low by sim
...
+
+### Owncast Config
+
+To point your owncast instance to the MinIO storage, configure the `s3` section as
+
+{{< highlight yaml >}}
+s3:
+  enabled: true
+  endpoint: https://<s3-storage.your.org>
+  servingEndpoint: https://<s3-storage.your.org>/stream
```
Comment (gabek) [inline content/docs/S3.md:224]: `servingEndpoint` is optional and only required if the hostname you want to serve from is different than the rl that the S3 provider returns.  I'm not sure if this is the case with MinIO or not.

File: content/docs/S3.md (hunk 5 above)
Comment (felix-engelmann) [reply content/docs/S3.md:214]: @jeyemwey maybe documenting the plain http headers migt be better?

File: content/docs/S3.md (hunk 6 above)
Comment (felix-engelmann) [reply content/docs/S3.md:224]: As minio does not support DNS subdomain style bucket addresses, we need the servingEndpoint to append the bucket name in the path.

**Comment by jeyemwey** [comment]
Looks good to me, thank you for your help and contributions!

**Review by jeyemwey — APPROVED** [review]

**Pull request merged**